File: uniffi_bindgen/src/bindings/swift/templates/CustomType.swift
{%- match config %}
{%- when None %}
{#- No config, just forward all methods to our builtin type #}
fileprivate typealias FfiConverterType{{ name }} = {{ builtin|ffi_converter_name }}

{%- when Some with (config) %}
fileprivate struct FfiConverterType{{ name }} {
    {#- Custom type config supplied, use it to convert the builtin type #}

    static func read(from buf: Reader) throws -> {{ self.type_name(config) }} {
        let builtinValue = try {{ builtin|read_fn }}(from: buf)
        return {{ config.into_custom.render("builtinValue") }}
    }

    static func write(_ value: {{ self.type_name(config) }}, into buf: Writer) {
        let builtinValue = {{ config.from_custom.render("value") }}
        return {{ builtin|write_fn }}(builtinValue, into: buf)
    }

    static func lift(_ value: {{ self.builtin_ffi_type().borrow()|type_ffi_lowered }}) throws -> {{ self.type_name(config) }} {
        let builtinValue = try {{ builtin|lift_fn }}(value)
        return {{ config.into_custom.render("builtinValue") }}
    }

    static func lower(_ value: {{ self.type_name(config) }}) -> {{ self.builtin_ffi_type().borrow()|type_ffi_lowered }} {
        let builtinValue = {{ config.from_custom.render("value") }}
        return {{ builtin|lower_fn }}(builtinValue)
    }

}
{%- endmatch %}

File: uniffi_bindgen/src/bindings/python/templates/CustomType.py
{%- let builtin = custom.builtin %}
{%- match custom.config %}
{% when None %}
{#- No custom type config, just forward all methods to our builtin type #}
class {{ custom|ffi_converter_name }}:
    @staticmethod
    def write(value, buf):
        {{ builtin|ffi_converter_name }}.write(value, buf)

    @staticmethod
    def read(buf):
        return {{ builtin|ffi_converter_name }}.read(buf)

    @staticmethod
    def lift(value):
        return {{ builtin|ffi_converter_name }}.lift(value)

    @staticmethod
    def check_lower(value):
        return {{ builtin|ffi_converter_name }}.check_lower(value)

    @staticmethod
    def lower(value):
        return {{ builtin|ffi_converter_name }}.lower(value)

{# Render a type alias from the custom type name to the concrete type name #}
{{ custom.name }} = {{ builtin|type_name -}}

{%- when Some(config) %}

{%- if let Some(type_name) = config.type_name %}
{# Render a type alias from the custom type name to the concrete type name #}
{{ custom.name }} = {{ type_name }}
{%- endif %}

{#- Custom type config supplied, use it to convert the builtin type #}
class {{ custom|ffi_converter_name }}:
    @staticmethod
    def write(value, buf):
        builtin_value = {{ config.lower("value") }}
        {{ builtin|write_fn }}(builtin_value, buf)

    @staticmethod
    def read(buf):
        builtin_value = {{ builtin|read_fn }}(buf)
        return {{ config.lift("builtin_value") }}

    @staticmethod
    def lift(value):
        builtin_value = {{ builtin|lift_fn }}(value)
        return {{ config.lift("builtin_value") }}

    @staticmethod
    def check_lower(value):
        builtin_value = {{ config.lower("value") }}
        return {{ builtin|check_lower_fn }}(builtin_value)

    @staticmethod
    def lower(value):
        builtin_value = {{ config.lower("value") }}
        return {{ builtin|lower_fn }}(builtin_value)
{%- endmatch %}
